### PR TITLE
Support null values for xgboost

### LINKED ISSUE #2888: Support None value in wide table for XGBoost train and explain.
In the Jupyter Notebook of the [Boston house tutorial](https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/tutorial/housing-explain.md), input the following SQL statements:

```SQL
CREATE TABLE boston.train_ext AS SELECT * FROM boston.train;

UPDATE boston.train_ext
SET rad = NULL
WHERE zn < 18.1 AND zn > 17.0;

SELECT * FROM boston.train_ext
TO TRAIN xgboost.gbtree
WITH
    objective="reg:squarederror",
    train.num_boost_round = 30
COLUMN crim, zn, indus, chas, nox, rm, age, dis, rad, tax, ptratio, b, lstat
LABEL medv
INTO sqlflow_models.my_xgb_regression_model;
```

We will get the following error messages:

```TXT
Traceback (most recent call last):

  File "<stdin>", line 108, in <module>

  File "/opt/sqlflow/python/runtime/xgboost/train.py", line 170, in train

    for per_batch_dmatrix in dtrain:

  File "/opt/sqlflow/python/runtime/xgboost/dataset.py", line 89, in xgb_dataset

    raw_data_dir=raw_data_dir)

  File "/opt/sqlflow/python/runtime/xgboost/dataset.py", line 130, in dump_dmatrix

    feature_metas)

  File "/opt/sqlflow/python/runtime/db.py", line 350, in read_features_from_row

    feature_metas[name], name)

  File "/opt/sqlflow/python/runtime/db.py", line 188, in read_feature

    int_raw_val = INT64_TYPE(raw_val)

TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

Since XGBoost model can handle the missing values, we can pass the `none` value into XGBoost directly instead of reporting error while dumping the dmatrix file.

### PR DESCRIPTION
**Pull request by typhoonzero**
Fix https://github.com/sql-machine-learning/sqlflow/issues/2888

### CODE REVIEW

File: python/runtime/db.py [hunk 1]
```diff
@@ -25,9 +25,13 @@ def connect_with_data_source(driver_dsn):
 
 
 INT64_TYPE = long if six.PY2 else int  # noqa: F821
+XGBOOST_NULL_MAGIC = 9999.0
 
 
 def read_feature(raw_val, feature_spec, feature_name):
+    # FIXME(typhoonzero): return None if not XGBoost
+    if raw_val is None:
+        return XGBOOST_NULL_MAGIC,
```
Comment (brightcoder01) [inline python/runtime/db.py]: What about the feature value of int type instead of double?
And how to handle np.array?

File: python/runtime/db.py [hunk 2]
```diff
@@ -25,50 +25,76 @@ def connect_with_data_source(driver_dsn):
...
+            else:
+                vec = np.fromstring(raw_val,
+                                    dtype=np.float32,
+                                    sep=feature_spec["delimiter"])
         elif feature_spec["dtype"] == "int64":
-            vec = np.fromstring(raw_val,
-                                dtype=np.int64,
-                                sep=feature_spec["delimiter"])
+            if raw_val is None:
+                vec = np.array([int(XGBOOST_NULL_MAGIC)], dtype=np.int64)
```
Comment (brightcoder01) [inline python/runtime/db.py]: The shape is not matched with the not-none case.
Will there be any offset inconsistency between none/not-none case inside [dump_matrix](https://github.com/sql-machine-learning/sqlflow/blob/c67113ecd3cc3115aac8bf56f0f7439054969de6/python/runtime/xgboost/dataset.py#L149-L155).

File: python/runtime/db.py [hunk 3]
```diff
@@ -25,50 +25,85 @@ def connect_with_data_source(driver_dsn):
...
+                                    dtype=np.float32,
+                                    sep=feature_spec["delimiter"])
         elif feature_spec["dtype"] == "int64":
-            vec = np.fromstring(raw_val,
-                                dtype=np.int64,
-                                sep=feature_spec["delimiter"])
+            if raw_val is None:
+                raise null_feature_error
+            else:
+                vec = np.fromstring(raw_val,
```
Comment (brightcoder01) [inline python/runtime/db.py:76]: How can we ensure that each item of a specific dense feature has the same shape?
np.fromstring may returns various shape for different input strings.

Comment (typhoonzero) [reply python/runtime/db.py:76]: This is done by feature derivation before training.

**Review by brightcoder01 — APPROVED** [review]

**Pull request merged**